lib/tree.rb

```diff
@@ -16,42 +16,141 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  #what to do if node is empty, smaller or larger than initial node
+  def add_helper(node, key, value)
+    #if the node is empty, the create the node with a key, value
+    if node.nil?
+      node = TreeNode.new(key, value)
+      #if the key is less than the node key, then put it to the left side
+    elsif key < node.key
+      node.left = add_helper(node.left, key, value)
+      #if key is larger than the node key, then add it to the right side
+    elsif key > node.key
+      node.right = add_helper(node.right, key, value)
+    end
+    return node
+  end
+
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
   def add(key, value)
```

Comment:
👍 The time/space complexities are O(log n) if the tree is balanced and O(n) if they are unbalanced.